src/main/java/org/broadinstitute/hellbender/utils/reference/FastaReferenceWriter.java

```diff
@@ -0,0 +1,665 @@
+        Utils.nonNull(bases, "the input bases cannot be negative");
+        ParamUtils.isPositiveOrZero(offset, "the input offset cannot be negative");
+        ParamUtils.isPositiveOrZero(length, "the input length must not be negative");
+        final int to = offset + length;
+        Utils.validateArg(to <= bases.length, "the length + offset goes beyond the end of " +
+                "the input base array: '" + to + "' > '" + bases.length + "'");
+
+        int next = offset;
+        while (next < to) {
+            if (currentLineBasesCount == currentBasesPerLine) {
+                fastaStream.write(LINE_SEPARATOR);
+                currentLineBasesCount = 0;
+            }
+            final int nextLength = Math.min(to - next, currentBasesPerLine - currentLineBasesCount);
+            fastaStream.write(bases, next, nextLength);
+            currentLineBasesCount += nextLength;
+            next += nextLength;
+        }
+        currentBasesCount += length;
+        return this;
+    }
+
+    private void assertSequenceOpen() {
+        if (currentSequenceName == null) {
+            throw new IllegalStateException("trying to add bases without starting a sequence");
+        }
+    }
+
+    private void assertIsNotClosed() {
+        if (closed) {
+            throw new IllegalStateException("already closed");
+        }
+    }
+
+    /**
+     * Closes this writer flushing all remaining writing operation input the output resources.
+     * <p>
+     *     Further calls to {@link #appendBases} or {@link #startSequence} will result in an exception.
+     * </p>
+     *
+     * @throws IOException if such exception is thrown when closing output writers and output streams.
+     */
+    public void close() throws IOException
+    {
+        if (!closed) {
+            closeSequence();
+            if (sequenceNamesAndSizes.isEmpty()) {
+                throw new IllegalStateException("no sequences where added to the reference");
+            }
+            fastaStream.close();
+            indexWriter.close();
+            if (dictOutput != null) {
+                outputDictionary();
+                dictOutput.close();
+            }
+            closed = true;
+        }
+    }
+
+    // print out the dictionary
+    private void outputDictionary() {
+        final SAMSequenceDictionary dictionary = new SAMSequenceDictionary(
+                sequenceNamesAndSizes.entrySet().stream()
+                    .map(e -> new SAMSequenceRecord(e.getKey(),  e.getValue().intValue()))
+                    .collect(Collectors.toList())
+        );
+        final SAMFileHeader header = new SAMFileHeader();
+        header.setSequenceDictionary(dictionary);
+
+        // use a empty try-resource results in compilation warnings due to the empty
+        // try body, so I do the traditional try-catch here.
+        try {
+            new SAMFileWriterFactory().makeSAMWriter(header, false, dictOutput).close();
```

Comment:
it seems awkward to use a SamFileWriter here, why not just use a SAMSequenceDictionaryCodec and encode them manually?

Reply:
Done.